wayland-backend/src/rs/server/mod.rs

```diff
@@ -0,0 +1,216 @@
+    /// Notification that the object has been destroyed and is no longer active
+    fn destroyed(&self, client_id: ClientId, object_id: ObjectId);
+    /// Helper for forwarding a Debug implementation of your `ObjectData` type
+    ///
+    /// By default will just print `ObjectData { ... }`
+    fn debug(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ObjectData").finish_non_exhaustive()
+    }
+}
+
+downcast_rs::impl_downcast!(sync ObjectData<D>);
+
+#[cfg(not(tarpaulin_include))]
+impl<D> std::fmt::Debug for dyn ObjectData<D> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.debug(f)
+    }
+}
+
+/// A trait representing the handling of new bound globals
+pub trait GlobalHandler<D>: downcast_rs::DowncastSync {
+    /// Check if given client is allowed to interact with given global
+    ///
+    /// If this function returns false, the client will not be notified of the existence
+    /// of this global, and any attempt to bind it will result in a protocol error as if
+    /// the global did not exist.
+    ///
+    /// Default implementation always return true.
+    fn can_view(
+        &self,
+        _client_id: ClientId,
+        _client_data: &Arc<dyn ClientData<D>>,
+        _global_id: GlobalId,
+    ) -> bool {
+        true
+    }
+    /// A global has been bound
+    ///
+    /// Given client bound given global, creating given object.
+    ///
+    /// The method must return the object data for the newly created object.
+    fn bind(
+        self: Arc<Self>,
+        handle: &mut Handle<D>,
+        data: &mut D,
+        client_id: ClientId,
+        global_id: GlobalId,
+        object_id: ObjectId,
+    ) -> Arc<dyn ObjectData<D>>;
+    /// Helper for forwarding a Debug implementation of your `GlobalHandler` type
+    ///
+    /// By default will just print `GlobalHandler { ... }`
+    fn debug(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GlobalHandler").finish_non_exhaustive()
+    }
+}
+
+downcast_rs::impl_downcast!(sync GlobalHandler<D>);
+
+/// A trait representing your data associated to a clientObjectData
+pub trait ClientData<D>: downcast_rs::DowncastSync {
+    /// Notification that a client was initialized
+    fn initialized(&self, client_id: ClientId);
+
+    /// Notification that a client is disconnected
+    fn disconnected(&self, client_id: ClientId, reason: DisconnectReason);
+    /// Helper for forwarding a Debug implementation of your `ClientData` type
+    ///
+    /// By default will just print `GlobalHandler { ... }`
+    fn debug(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ClientData").finish_non_exhaustive()
+    }
+}
+
+downcast_rs::impl_downcast!(sync ClientData<D>);
+
+/// An id of an object on a wayland server.
+#[derive(Clone)]
+pub struct ObjectId {
+    id: u32,
+    serial: u32,
+    client_id: ClientId,
```

Comment:
Could we add a getter for `client_id` here?
Or directly in `Resource` trait.

I basically need a way to replicate the behavior of `same_client_as`, so I need a way to get `ClientId`.

Reply:
Would a `same_client_as` method on `ObjectId` do the job, instead? Producing a valid `ClientId` without the `Handle` is not going to be possible on the `sys` backend.

Reply:
I haven't needed direct access to client_Id so far, so just a way to replicate `same_client_as` behavior should be enough.

Reply:
Just as a sidenote, while at it, would it be somehow possible to avoid it's current behavior?
```rs
if !(self.is_alive() && other.is_alive()) {
       return false;
}
 ```
 Pattern of cleaning up stuff in Smithay by using `same_client_as` inside destructors is all over the place, and from my testing I'm pretty sure it does not clean up properly because objects have to be alive to compare them.